### PR TITLE
Feature: puback outbound/inbound interceptor

### LINKED ISSUE #50: Implement PubackOutboundInterceptor

### PR DESCRIPTION
**Pull request by YannickWeber**
Resolves #50 and #49 

**Motivation**
Resolves the absence of a possibility for extension developers to intercept outbound puback messages.

**Changes**
Implemented the puback outbound/intbound interceptors.
An extension-developer is now able to intercept outbound/inbound PUBACK messages.
The developer can change the following parameters of the PUBACK message:
* AckReasonCode (changing from a successful to an unsuccessful reason code is prohibited and vice versa)
* ReasonString (can only be set for an unsuccessful PUBACK message)
* when an exception is caught in the interceptor, the client is not disconnected

Exceptions / Timeouts are handled in the following way:
* when an exception is caught in the interceptor, the client is not disconnected
* when a timeout fails in the interceptor, the client is not disconnected
* when an exception is caught in the interceptor all changes to the PUBREC are ignored.

### CODE REVIEW

**Comment by MagiErber** [comment]
@cla-bot check

File: hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/client/ClientContext.java [hunk 1]
```diff
@@ -136,6 +155,14 @@
     @Immutable
     @NotNull List<@NotNull SubscribeInboundInterceptor> getSubscribeInboundInterceptors();
 
+    /**
+     * Returns all {@link PubackOutboundInterceptor} which are registered for this client by this extension.
+     *
+     * @return List of {@link PublishInboundInterceptor} for this client.
```
Comment (sauroter) [inline hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/client/ClientContext.java]:
```suggestion
     * @return List of {@link PubackOutboundInterceptor} for this client.
```

File: hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/packets/puback/ModifiablePubackPacket.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * Set a {@link AckReasonCode} to the PUBACK packet.
+     * <p>
+     * Switching from successful to unsuccessful and vice versa is not supported.
+     *
+     * @param reasonCode The reason code to set.
+     * @throws NullPointerException  If reason code is <null>.
+     * @throws IllegalStateException If switching from successful reason code to unsuccessful reason code or vice versa.
+     * @see AckReasonCode How reason codes are translated from MQTT 5 to MQTT 3.
+     */
+    void setReasonCode(final @NotNull AckReasonCode reasonCode);
```
Comment (sauroter) [inline hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/packets/puback/ModifiablePubackPacket.java]: I dont like AckReasonCode. That sounds bad. Perhapps make it abstract and add an PubackReasonCode?

File: hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/packets/puback/ModifiablePubackPacket.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * A {@link PubackPacket} that can be modified before it is sent to the client.
+ *
+ * @author Yannick Weber
+ */
+public interface ModifiablePubackPacket extends PubackPacket {
+
+    /**
+     * Set a {@link AckReasonCode} to the PUBACK packet.
+     * <p>
+     * Switching from successful to unsuccessful and vice versa is not supported.
```
Comment (sauroter) [inline hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/packets/puback/ModifiablePubackPacket.java]: What happens if i still do it?

File: src/main/java/com/hivemq/bootstrap/netty/ChannelDependencies.java [hunk 4]
```diff
@@ -164,6 +165,7 @@
 
     @NotNull
     private final ConnackOutboundInterceptorHandler connackOutboundInterceptorHandler;
+    private final @NotNull PubackOutboundInterceptorHandler pubackOutboundInterceptorHandler;
```
Comment (sauroter) [inline src/main/java/com/hivemq/bootstrap/netty/ChannelDependencies.java]: Please unify nn and nm placements.

File: hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/packets/puback/ModifiablePubackPacket.java (hunk 3 above)
Comment (YannickWeber) [reply hivemq-extension-sdk/src/main/java/com/hivemq/extension/sdk/api/packets/puback/ModifiablePubackPacket.java]: @throws IllegalStateException If switching from successful reason code to unsuccessful reason code or vice versa.

I should add this to the documentation text.

File: src/main/java/com/hivemq/extensions/handler/PubackOutboundInterceptorHandler.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        private final @NotNull PubackOutboundOutputImpl output;
+        private final @NotNull SettableFuture<Void> interceptorFuture;
+        private final int interceptorCount;
+        private final @NotNull AtomicInteger counter;
+
+
+        PubackInterceptorContext(final @NotNull Class<?> taskClazz,
+                                  final @NotNull String clientId,
+                                  final @NotNull PubackOutboundInputImpl input,
+                final @NotNull PubackOutboundOutputImpl output,
```
Comment (sauroter) [inline src/main/java/com/hivemq/extensions/handler/PubackOutboundInterceptorHandler.java]: Weird formatting here

File: src/main/java/com/hivemq/extensions/interceptor/puback/PubackOutboundOutputImpl.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    public PubackOutboundOutputImpl get() {
+        return this;
+    }
+
+    public void update(final @NotNull PUBACK puback) {
+        modifiablePubackPacket = new ModifiablePubackPacketImpl(configurationService, puback);
+    }
+
+    public void update(final @NotNull ModifiablePubackPacketImpl packet) {
+        this.modifiablePubackPacket = packet;
```
Comment (sauroter) [inline src/main/java/com/hivemq/extensions/interceptor/puback/PubackOutboundOutputImpl.java]: Isnt here a deep copy neccessary, as we give out the object to the extensions.

File: src/main/java/com/hivemq/extensions/handler/PubackOutboundInterceptorHandler.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+        final PUBACK puback = (PUBACK) msg;
+
+        final Channel channel = ctx.channel();
+        if (!channel.isActive()) {
+            return;
+        }
+
+        final String clientId = channel.attr(ChannelAttributes.CLIENT_ID).get();
+        if (clientId == null) {
```
Comment (sauroter) [inline src/main/java/com/hivemq/extensions/handler/PubackOutboundInterceptorHandler.java]: Why is the channel promise not failed?

**Pull request closed without merging**